Support registration of external backends

mrocklin opened the pull request:
Fixes https://github.com/joblib/joblib/issues/653

Currently when using external backends with joblib you need to know to
`import distributed.joblib`, which is not particularly intuitive.

```python
from joblib import parallel_backend
with parallel_backend("dask"):
    pass

KeyError: 'dask'
```

Now implementers of downstream parallel backends can register small
functions within the Joblib codebase that register their backends
appropriately when triggered by users.

FWIW I suspect that not everyone will agree with this change.  I will not be surprised if it does not get merged.  It was simple enough to implement though that I thought I'd throw it up here just to make conversation about it easier.

ogrisel commented:
I was thinking that maybe we could even move the code of the dask backend into the joblib code base because the backend semi-private API is more subject to change that the dask distributed public API.

I would like to find some time in the coming month to automatically scatter heavy data to avoid redundant data transfer without having the use to explicitly use the `scatter` feature of the backend. This refactoring would be more natural to do as part of the joblib code base I think.

GaelVaroquaux commented:
This code is fairly lightweight, and it is gives benefit to the user, so I have no major issues with the idea.

Two points:

* We should somewhat briefly document this, so that other backend providers can use it. The following page would be a good one: https://joblib.readthedocs.io/en/latest/parallel.html#custom-backend-api-experimental

* Do we want to fail gracefully when distributed is no available? Maybe that's for future work, to be able to specify fallbacks while choosing the backend.

mrocklin commented:
Those both seem reasonable to fix.  

> Do we want to fail gracefully when distributed is no available? Maybe that's for future work, to be able to specify fallbacks while choosing the backend.

I agree that this would be useful.

Short-term I've added a more informative error.

> We should somewhat briefly document this, so that other backend providers can use it. The following page would be a good one: https://joblib.readthedocs.io/en/latest/parallel.html#custom-backend-api-experimental

Done.  Feedback appreciated to check for tone and and detail.

I've also changed the dask example.

GaelVaroquaux commented:
LGTM. Merging!

GaelVaroquaux commented:
Actually, let me wait for travis

mrocklin commented:
Tests pass except for codecov.  I'm not sure how to interpret the results there.
